src/dmd/root/filename.d

```diff
@@ -856,8 +866,45 @@ nothrow:
     {
         version (Posix)
         {
-            // NULL destination buffer is allowed and preferred
-            return name.toCStringThen!((n) => realpath(n.ptr, null)).toDString;
+            import core.stdc.limits;      // PATH_MAX
+            import core.sys.posix.unistd; // _PC_PATH_MAX
+
+            // Have realpath(), passing a NULL destination pointer may return an
+            // internally malloc'd buffer, however it is implementation defined
+            // as to what happens, so cannot rely on it.
+            static if (__traits(compiles, PATH_MAX))
+            {
+                // Have compile time limit on filesystem path, use it with realpath.
+                char[PATH_MAX] buf = void;
+                auto path = name.toCStringThen!((n) => realpath(n.ptr, buf.ptr));
+                if (path !is null)
+                    return mem.xstrdup(path).toDString;
+            }
+            else static if (__traits(compiles, canonicalize_file_name))
+            {
+                // Have canonicalize_file_name, which malloc's memory.
+                auto path = name.toCStringThen!((n) => canonicalize_file_name(n.ptr));
+                if (path !is null)
+                    return path.toDString;
+            }
+            else static if (__traits(compiles, _PC_PATH_MAX))
+            {
+                // Panic! Query the OS for the buffer limit.
+                auto path_max = pathconf("/", _PC_PATH_MAX);
+                if (path_max > 0)
+                {
+                    char *buf = cast(char*)mem.xmalloc(path_max);
+                    scope(exit) mem.xfree(buf);
+                    auto path = name.toCStringThen!((n) => realpath(n.ptr, buf));
+                    if (path !is null)
+                        return mem.xstrdup(path).toDString;
+                }
+            }
+            // Give up trying to support this platform, just duplicate the filename
+            // unless there is nothing to copy from.
+            if (!name.length)
+                return null;
+            return mem.xstrdup(name.ptr)[0 .. name.length];
```

Comment:
Why not a `static assert` instead of doing nothing ? It does not look like a good default behavior and might be hard to track.

Reply:
Doing nothing would mean nothing would compile. We already have the relative paths to druntime and phobos, let the compiler use them instead of erroring that it can't find any sources.